server/app/services/pubsub/base.rb

```diff
@@ -0,0 +1,139 @@
+require_relative '../logging'
+
+module Pubsub
+  class Base
+    include Celluloid
+    include Logging
+
+    class Subscription
+      include Logging
+
+      attr_reader :channel
+
+      # @param [String] channel
+      # @param [Proc] block
+      def initialize(channel, block)
+        @channel = channel
+        @block = block
+        @queue = Queue.new
+      end
+
+      def terminate
+        stop
+      end
+
+      def stop
+        @queue.close
+      end
+
+      def queue_message(data)
+        @queue << data
+      rescue ClosedQueueError
+        nil
+      end
+
+      # returns once stopped, and queued messages have been processed
+      def process
+        while data = @queue.shift
+          send_message(data)
+        end
+      end
+
+      private
+
+      # @param [Hash] data
+      def send_message(data)
+        payload = HashWithIndifferentAccess.new(data)
```

Comment:
Changed this a bit. Previously the passed data was not actually a hash (it was `BSON::Binary`)... it seemed like a bug because it also unpacked binary for each subscriber. Now unpack happens in mongo/redis backends.